src/test/librbd/io/test_mock_ImageRequestWQ.cc

```diff
@@ -306,5 +327,29 @@ TEST_F(TestMockIoImageRequestWQ, RefreshError) {
   aio_comp->release();
 }
 
+TEST_F(TestMockIoImageRequestWQ, IOPSQos) {
+  librbd::ImageCtx *ictx;
+  ASSERT_EQ(0, open_image(m_image_name, &ictx));
+
+  MockTestImageCtx mock_image_ctx(*ictx);
+
+  MockImageDispatchSpec mock_queued_image_request;
+  expect_was_throttled(mock_queued_image_request, false);
+  expect_set_throttled(mock_queued_image_request);
+
+  InSequence seq;
+  MockImageRequestWQ mock_image_request_wq(&mock_image_ctx, "io", 60, nullptr);
+
+  mock_image_request_wq.apply_qos_limit(1, RBD_QOS_IOPS_THROTTLE);
```

Comment:
I haven't run it myself to confirm, but is this test showing that we don't initially populate the token bucket upon startup (i.e. this test will pause for 1 second)? If that's the case, perhaps we should always initially populate the bucket so that QoS user's don't always have a 1 second IO pause after opening the image.

Reply:
Hi @dillaman good point. I did a test for this:
```
[root@atest-guest build]# rbd bench --io-type read --io-size 4096 --io-threads 1 test1
bench  type read io_size 4096 io_threads 1 bytes 1073741824 pattern sequential
  SEC       OPS   OPS/SEC   BYTES/SEC
    1        20     21.02  86101.18
    2        40     20.50  83967.10
    3        60     20.33  83284.44
    4        80     20.25  82943.11
^C
[root@atest-guest build]# rbd image-meta list test1
There is 1 metadatum on this image:

Key                     Value 
conf_rbd_qos_iops_limit 20   
```
The result did not show the pause of 1 second after opening the image. But I agree that we need to initially populate the token bucket. There are two cases we will init a token. 

(1) constructer of Bucket:
```
    Bucket(CephContext *cct, const std::string& n, uint64_t m)
      : cct(cct), name(n),
        remain(m), max(m)
    {
    }
```
That's okey, we initialized the remain with m which means the capacity. 
(2) set_max()
```
void TokenBucketThrottle::Bucket::set_max(uint64_t m) {
  if (remain > m)
    remain = m;
  max = m;
}
```
There is a possibility in this case that when we change the state from non-limit to limit of 100, we only change the max from 0 to 100, but the remain is still 0 and wait for timer to put tokens. So I pushed a commit to fix this case.